typescript/sdk/src/ism/metadata/routing.ts

```diff
@@ -66,71 +70,63 @@ export class RoutingMetadataBuilder implements MetadataBuilder {
   }
 }
 
-export class DefaultFallbackRoutingMetadataBuilder extends RoutingMetadataBuilder {
+export class DynamicRoutingMetadataBuilder extends StaticRoutingMetadataBuilder {
+  constructor(protected baseMetadataBuilder: BaseMetadataBuilder) {
+    super(baseMetadataBuilder);
+  }
+
   public async build(
     context: MetadataContext<WithAddress<RoutingIsmConfig>>,
     maxDepth = 10,
   ): Promise<string> {
+    const { message, ism } = context;
     const originChain = this.baseMetadataBuilder.multiProvider.getChainName(
-      context.message.parsed.origin,
+      message.parsed.origin,
     );
+    const destination = message.parsed.destination;
+    const provider =
+      this.baseMetadataBuilder.multiProvider.getProvider(destination);
+
+    // Helper to derive new ISM config and recurse
+    const deriveAndRecurse = async (moduleAddress: string) => {
+      const ismReader = new EvmIsmReader(
+        this.baseMetadataBuilder.multiProvider,
+        destination,
+      );
+      const nextConfig = await ismReader.deriveIsmConfig(moduleAddress);
+      return this.baseMetadataBuilder.build(
+        { ...context, ism: nextConfig },
+        maxDepth - 1,
+      );
+    };
 
-    const isRouted =
-      context.ism.type === IsmType.AMOUNT_ROUTING
-        ? false
-        : !!context.ism.domains[originChain];
-    // If the chain is routed then we are 100% sure that the ism is not an ICA ISM
-    if (isRouted) {
+    // 1) Dynamic routing (AmountRouting or ICA): always route via .route(message)
+    if (isDynamicallyRoutedIsmType(ism.type)) {
+      const router = IRoutingIsm__factory.connect(ism.address, provider);
+      const moduleAddr = await router.route(message.message);
+      return deriveAndRecurse(moduleAddr);
+    }
```

Comment:
does it not seem odd to you that we are implementing ISM derivation in here rather than in the ISM reader?

Reply:
Lets discuss sync, but no, like we discussed, i dont think we have a strong practice of "doing max ISM derivation outside metadata building"

Reply:
Chatted offline, ideally the metadata builder can rely on ISM derivation having fully derived the ISM with the messageContext, but because it's not hard-enforced, we could end up in a situation where a caller derived the ISM without message context which would then require the metadata builder to do this derivation